File: clojure/space-age/src/space_age.clj
(ns space-age)

(def secs-in-earth-year (* 365.25 24 60 60))

(defn- days-in-year [period]
  (* secs-in-earth-year period))

(defn- on-planet [period seconds]
  (/ seconds (days-in-year period)))

(defn on-earth [seconds] (on-planet 1.0 seconds))
(defn on-mercury [seconds] (on-planet 0.2408467 seconds))
(defn on-venus [seconds] (on-planet 0.61519726 seconds))
(defn on-mars [seconds] (on-planet 1.8808158 seconds))
(defn on-jupiter [seconds] (on-planet 11.862615 seconds))
(defn on-saturn [seconds] (on-planet 29.447498 seconds))
(defn on-uranus [seconds] (on-planet 84.016846 seconds))
(defn on-neptune [seconds] (on-planet 164.79132 seconds))

File: haskell/space-age/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury | Venus | Earth | Mars | Saturn | Jupiter | Neptune | Uranus
  deriving (Show)

earthYearInSeconds :: Double
earthYearInSeconds = 365.25 * 24 * 60 * 60

ageOn :: Planet -> Integer -> Double
ageOn p secs =  (fromIntegral secs) / (earthYearInSeconds * (yearLength p))

yearLength Earth = 1.0
yearLength Mercury = 0.2408467
yearLength Venus = 0.61519726
yearLength Mars = 1.8808158
yearLength Jupiter = 11.862615
yearLength Saturn = 29.447498
yearLength Uranus = 84.016846
yearLength Neptune = 164.79132
